src/crypto/CMakeLists.txt

```diff
@@ -64,61 +66,99 @@ endif()
 include(CheckCXXSymbolExists)
 check_cxx_symbol_exists(timingsafe_bcmp "string.h" HAVE_TIMINGSAFE_BCMP)
 
-# The `bitcoin_crypto` must be of an `OBJECT` type to include
-# its object files into the output static library archives.
-add_library(bitcoin_crypto OBJECT EXCLUDE_FROM_ALL
-  $<$<BOOL:${ASM}>:sha256_sse4.cpp>
-  aes.cpp
-  chacha20.cpp
-  chacha20poly1305.cpp
-  hkdf_sha256_32.cpp
-  hmac_sha256.cpp
-  hmac_sha512.cpp
-  poly1305.cpp
-  muhash.cpp
-  ripemd160.cpp
-  sha1.cpp
-  sha256.cpp
-  sha3.cpp
-  sha512.cpp
-  siphash.cpp
-)
-target_compile_definitions(bitcoin_crypto
-  PRIVATE
-    $<$<BOOL:${ASM}>:USE_ASM>
-    $<$<BOOL:${HAVE_SSE41}>:ENABLE_SSE41>
-    $<$<BOOL:${HAVE_AVX2}>:ENABLE_AVX2>
-    $<$<BOOL:${HAVE_X86_SHANI}>:ENABLE_X86_SHANI>
-    $<$<BOOL:${HAVE_ARM_SHANI}>:ENABLE_ARM_SHANI>
-    $<$<BOOL:${HAVE_TIMINGSAFE_BCMP}>:HAVE_TIMINGSAFE_BCMP>
-)
-
 if(HAVE_SSE41)
-  target_sources(bitcoin_crypto PRIVATE sha256_sse41.cpp)
-  set_property(SOURCE sha256_sse41.cpp
-    APPEND PROPERTY COMPILE_OPTIONS ${SSE41_CXXFLAGS}
+  add_library(bitcoin_crypto_sse41 OBJECT EXCLUDE_FROM_ALL
+    sha256_sse41.cpp
+  )
+  target_compile_definitions(bitcoin_crypto_sse41
+    PUBLIC
+      ENABLE_SSE41
```

Comment:
Hmm, this all looks like it might be more complicated than necessary.

Upstream these are kinda a mess of autoconf/automake makefile preprocessors and c++ preprocessor defines. I wonder if it would be simpler if we tweaked the way these are defined first. I'll mess with that a little bit.

Reply:
> I wonder if it would be simpler if we tweaked the way these are defined first.

Any hints about the way how it could be "tweaked"?

Despite the suggest approach looks verbose, it follows the CMake's way :)

Everything is abstracted in targets, and we manipulate with targets' properties.

There are alternative approaches, though.

First approach can include `sha256_<*>.cpp` sources conditionally, and modify source's properties to add proper compiler flags.

Second one can apply feature-specific compiler flags for the whole `bitcoin_crypto` target. That's Makefile's way :) I'd like to avoid it.